packages/contract/src/pay/DaimoPay.sol

```diff
@@ -0,0 +1,269 @@
+import "./PayIntent.sol";
+import "./PayIntentFactory.sol";
+import "./TokenUtils.sol";
+
+// A Daimo Pay transfer has 4 steps:
+// 1. Alice sends (tokenIn, amountIn) to handoff address on chain A -- simple erc20 transfer
+// 2. LP swaps tokenIn to bridgeTokenIn and burns on chain A -- LP runs this in sendAndSelfDestruct
+//    - LP doesnt need approve or anything else, just quote and swap.
+//    - quote comes from the handoff address which commits to the destination bridgeTokenOut amount, and therefore bridgeTokenIn amount.
+//    - LP has to fetch swap from Uniswap or similar
+
+// Fork: fastFinish, then claim
+// Fork: claim directly
+
+// 4. LP swaps bridgeTokenOut to tokenOut on chain B -- LP runs this in completeAction
+// Alice is responsible for putting a quote for the bridgeTokenOut <> tokenOut swap
+// This fixes bridgeTokenOut expected amount, which in turn fixes the bridgeTokenIn burn amount,
+// locking in the amounts expected for all intermediary swaps.
+
+// To be nice Alice can put a a slightly worse quote than the market price to incentive LP.
+
+/// @title Daimo Pay contract for creating and fulfilling cross-chain actions
+/// @author The Daimo team
+/// @custom:security-contact security@daimo.com
+///
+/// Allows optimistic fast actions. Alice initiates a transfer by calling
+/// `startAction` on chain A. After the bridging delay (10+ min for CCTP),
+/// funds arrive at the intent address deployed on chain B. Bob can call
+/// `claimAction` function to perform Bob's action. Alternatively, immediately
+/// after the first call, an LP can call `fastFinishAction` to perform Bob's
+/// action immediately. Later, when the funds arrive from the bridge, the LP
+/// (rather than Bob) will claim.
+contract DaimoPay {
+    using SafeERC20 for IERC20;
+
+    PayIntentFactory public immutable intentFactory;
+    DaimoPayBridger public immutable bridger;
+
+    /// Commit to transfer details in an intent address.
+    mapping(address intentAddr => bool) public intentSent;
+    /// On the receiving chain, map each intent to a recipient (LP or Bob).
+    mapping(address intentAddr => address) public intentToRecipient;
+
+    // Action initiated on chain A
+    event Start(address indexed intentAddr, PayIntent intent);
+
+    // Action completed ~immediately on chain B
+    event FastFinish(
+        address indexed intentAddr,
+        address indexed newRecipient,
+        PayIntent intent
+    );
+
+    // Action settled later, once the underlying bridge transfer completes.
+    event Claim(
+        address indexed intentAddr,
+        address indexed finalRecipient,
+        PayIntent intent
+    );
+
+    // When the action is completed as expected, emit this event
+    event ActionCompleted(
+        address indexed intentAddr,
+        address indexed destinationAddress
+    );
+
+    // When the action is a call that fails, we bounce the funds to the
+    // specified refund address and emit this event
+    event ActionBounced(
```

Comment:
maybe ActionRefunded

Reply:
one more idea: we could delete these two events + add a status to `Claim`/`FastFinish`

Reply:
or combine into `IntentFinished` with a `boolean success`